File: 09-backtracking/00017-Letter_Combinations_Of_A_Phone_Number/00017-letter_combinations_of_a_phone_number.cpp
// 17: Letter Combinations Of A Phone Number
// https://leetcode.com/problems/letter-combinations-of-a-phone-number


#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
private:
    void backtrack(string digits, int index, unordered_map<char, string>& m, string& current, vector<string>& result) {
        if (index == digits.size()) {
            result.push_back(current);
            return;
        }

        string s = m[digits[index]];
        for (int i=0; i<s.size(); i++) {
            current.push_back(s[i]);
            backtrack(digits, index + 1, m, current, result);
            current.pop_back();
        }
    }


public:
    // SOLUTOIN
   vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return {};
        }
        
        unordered_map<char, string> m = {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}
        };
        
        string current;
        vector<string> result;
        backtrack(digits, 0, m, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    string digits = "23";

    // OUTPUT
    auto result = o.letterCombinations(digits);

    cout << "["; for (auto x : result) {
        cout << "\"" << x << "\", ";
    } cout << "\b\b]" << endl;

    return 0;
}

File: 09-backtracking/00017-Letter_Combinations_Of_A_Phone_Number/00017-letter_combinations_of_a_phone_number.go
// 17: Letter Combinations Of A Phone Number
// https://leetcode.com/problems/letter-combinations-of-a-phone-number

package main

import "fmt"

func backtrack(digits string, index int, m map[byte]string, current string, result *[]string) {
	if index == len(digits) {
		*result = append(*result, current)
		return
	}

	s := m[digits[index]]
	for i := 0; i < len(s); i++ {
		current += string(s[i])
		backtrack(digits, index+1, m, current, result)
		current = current[:len(current)-1]
	}
}

// SOLUTION
func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}

	m := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}

	var current string
	var result []string
	backtrack(digits, 0, m, current, &result)
	return result
}

func main() {
	// INPUT
	digits := "23"

	// OUTPUT
	result := letterCombinations(digits)

	fmt.Println(result)
}
